open_spiel/games/solitaire.h

```diff
@@ -0,0 +1,409 @@
+  SuitType GetSuit() const;
+  LocationType GetLocation() const;
+  bool GetHidden() const;
+  int GetIndex() const;
+
+  // Setters
+  void SetRank(RankType new_rank);
+  void SetSuit(SuitType new_suit);
+  void SetLocation(LocationType new_location);
+  void SetHidden(bool new_hidden);
+
+  // Operators
+  bool operator==(const Card &other_card) const;
+  bool operator<(const Card &other_card) const;
+
+  // Other Methods
+  std::string ToString(bool colored = true) const;
+  std::vector<Card> LegalChildren() const;
+};
+
+class Pile {
+ protected:
+  std::vector<Card> cards;
+  const LocationType type;
+  const SuitType suit;
+  const PileID id;
+  const int max_size;
+
+ public:
+  // Constructor
+  Pile(LocationType type, PileID id, SuitType suit = kSuitNone);
+
+  // Destructor
+  virtual ~Pile() = default;
+
+  // Getters/Setters
+  bool GetIsEmpty() const;
+  SuitType GetSuit() const;
+  LocationType GetType() const;
+  PileID GetID() const;
+  Card GetFirstCard() const;
+  Card GetLastCard() const;
+  std::vector<Card> GetCards() const;
+  void SetCards(std::vector<Card> new_cards);
+
+  // Other Methods
+  virtual std::vector<Card> Sources() const;
+  virtual std::vector<Card> Targets() const;
+  virtual std::vector<Card> Split(Card card);
+  virtual void Reveal(Card card_to_reveal);
+  void Extend(std::vector<Card> source_cards);
+  std::string ToString(bool colored = true) const;
+};
+
+class Tableau : public Pile {
+ public:
+  // Constructor
+  explicit Tableau(PileID id);
+
+  // Other Methods
+  std::vector<Card> Sources() const override;
+  std::vector<Card> Targets() const override;
+  std::vector<Card> Split(Card card) override;
+  void Reveal(Card card_to_reveal) override;
+};
+
+class Foundation : public Pile {
+ public:
+  // Constructor
+  Foundation(PileID id, SuitType suit);
+
+  // Other Methods
+  std::vector<Card> Sources() const override;
+  std::vector<Card> Targets() const override;
+  std::vector<Card> Split(Card card) override;
+};
+
+class Waste : public Pile {
+ public:
+  // Constructor
+  Waste();
+
+  // Other Methods
+  std::vector<Card> Sources() const override;
+  std::vector<Card> Targets() const override;
+  std::vector<Card> Split(Card card) override;
+  void Reveal(Card card_to_reveal) override;
+};
+
+class Move {
+ private:
+  Card target;
+  Card source;
+
+ public:
+  // Constructors
+  Move(Card target_card, Card source_card);
+  Move(RankType target_rank, SuitType target_suit, RankType source_rank,
+       SuitType source_suit);
+  explicit Move(Action action);
+
+  // Getters
+  Card GetTarget() const;
+  Card GetSource() const;
+
+  // Other Methods
+  // ===========================================================================
+  std::string ToString(bool colored = true) const;
+  bool operator<(const Move &other_move) const;
+  Action ActionId() const;
+};
+
+// OpenSpiel Classes ===========================================================
+
+class SolitaireGame;
```

Comment:
I think we can move the declaration for SolitaireGame to come before SolitaireState, and then we can remove the forward declaration.